### PR TITLE
♻️ Union list support

### PR DESCRIPTION
**Pull request by prashantgupta24**
Add example to support simple lists and lists in Union!

Heavy lifting done by this PR: https://github.com/IBM/py-to-proto/pull/54

fix https://github.com/caikit/caikit/issues/198

(also please squash commits, unable to git squash when there's a `merge main` commit)

### CODE REVIEW

File: caikit/runtime/utils/servicer_util.py [hunk 1]
```diff
@@ -47,8 +47,15 @@ def validate_caikit_library_class_exists(cdm, class_name):
         return getattr(cdm, class_name)
 
     except AttributeError as e:
-        log.error("<RUN24024010E>", "No Caikit Library CDM class for %s", class_name)
-        raise e
+        log.warning("<RUN24024050W>", "No Caikit Library CDM class for %s", class_name)
```
Comment (prashantgupta24) [inline caikit/runtime/utils/servicer_util.py:50]: Based on internal discussions, I'm also okay in removing the whole `validate_data_model` function altogether!

Comment (joerunde) [reply caikit/runtime/utils/servicer_util.py:51]: Here's a story: https://github.com/caikit/caikit/issues/247

File: tests/core/data_model/test_dataobject.py [hunk 2]
```diff
@@ -878,6 +878,51 @@ class Derived(Base):
...
+
+    #     @dataobject
+    #     class StrSequence(DataObjectBase):
+    #         values: List[str]
+
+    #     foo: Union[IntSequence, StrSequence]
+
+    # proto round trip
+    foo_int = Foo.IntSequence(values=[1, 2])
+    foo1 = Foo(foo=foo_int)
```
Comment (joerunde) [inline tests/core/data_model/test_dataobject.py:905]: Is this the only way these can be constructed?

It's good, but I think a native python @dataclass user would expect to be able to write
```
foo_with_ints = Foo([1, 2, 3])
foo_with_strs = Foo(["1", "2", "3"])
```

File: tests/runtime/utils/test_servicer_util.py [hunk 3]
```diff
@@ -68,7 +68,7 @@ def test_servicer_util_validate_caikit_library_class_exists_returns_caikit_class
 
 def test_servicer_util_validate_caikit_library_class_exists_raises_for_garbage_input():
     """Test that validate_caikit_library_class_exists raises for garbage 'model'"""
-    with pytest.raises(AttributeError):
+    with pytest.raises(ValueError):
```
Comment (joerunde) [inline tests/runtime/utils/test_servicer_util.py:71]: I don't understand how your changes made this now be a value error? 🤷 

But I'm okay with it if we're discussing removing these checks anyway

File: tests/fixtures/sample_lib/modules/other_task/other_implementation.py [hunk 4]
```diff
@@ -51,11 +51,13 @@ def train(
         cls,
         training_data: DataStream[int],
         sample_input: Union[SampleInputType, str],
+        list_value: Union[List[str], List[int]],
```
Comment (joerunde) [inline tests/fixtures/sample_lib/modules/other_task/other_implementation.py:54]: I like the completeness of testing, but I do worry that we don't have much organization of these test fixtures.
There is a "primitive party implementation" test module, should we put all of these python primitive cases there instead?

**Review by joerunde — CHANGES_REQUESTED** [review]
I probably missed the design discussion around how to instantiate these things in python, did we decide to not support simple lists or is that just not in a test? (Or did I just miss it?)

**Comment by joerunde** [comment]
Yeah I don't know about the git history problem, but I'm fine if we just use the squash merge button when merging if you want

File: tests/fixtures/sample_lib/modules/other_task/other_implementation.py (hunk 4 above)
Comment (prashantgupta24) [reply tests/fixtures/sample_lib/modules/other_task/other_implementation.py:54]: Can we open a separate issue on organization of test fixtures and tackle it there?

Comment (joerunde) [reply tests/fixtures/sample_lib/modules/other_task/other_implementation.py:54]: > Actually let me change that in this PR itself

Yup, I'd rather do it this way. Thanks!

Should this PR wait to merge until the dict PR merges?

**Comment by prashantgupta24** [comment]
> did we decide to not support simple lists or is that just not in a test? (Or did I just miss it?)

I can add a test for that, I'm assuming the only thing missing is a test for it

File: tests/core/data_model/test_dataobject.py (hunk 2 above)
Comment (prashantgupta24) [reply tests/core/data_model/test_dataobject.py:905]: Hmm, yeah that would be ideal, but can that be a separate feature request? This PR is mostly to handle the lists within function arguments, I believe in order to support the above, there _might_ have to be metaclass changes? Not sure

File: caikit/runtime/utils/servicer_util.py (hunk 1 above)
Comment (prashantgupta24) [reply caikit/runtime/utils/servicer_util.py:51]: Will handle in subsequent PR

**Comment by prashantgupta24** [comment]
Hey @joerunde, update on this PR:

- added the test for a simple list
- moved test to `primitive_party`
- separate issue for removing validation of data model
- separate issue for python style initialization of lists within unions

**Review by joerunde — APPROVED** [review]
lpgtm!

**Pull request merged**